horovod/spark/common/store.py

```diff
@@ -0,0 +1,338 @@
+
+    def get_val_data_path(self, idx=None):
+        raise NotImplementedError()
+
+    def get_test_data_path(self, idx=None):
+        raise NotImplementedError()
+
+    def saving_runs(self):
+        raise NotImplementedError()
+
+    def get_runs_path(self):
+        raise NotImplementedError()
+
+    def get_run_path(self, run_id):
+        raise NotImplementedError()
+
+    def get_checkpoint_path(self, run_id):
+        raise NotImplementedError()
+
+    def get_logs_path(self, run_id):
+        raise NotImplementedError()
+
+    def get_checkpoint_filename(self):
+        raise NotImplementedError()
+
+    def get_logs_subdir(self):
+        raise NotImplementedError()
+
+    def exists(self, path):
+        raise NotImplementedError()
+
+    def read(self, path):
+        raise NotImplementedError()
+
+    def get_local_output_dir_fn(self, run_id):
+        raise NotImplementedError()
+
+    def sync_fn(self, run_id):
+        raise NotImplementedError()
+
+    def to_remote(self, run_id, dataset_idx):
+        attrs = self._remote_attrs(run_id, dataset_idx)
+
+        class RemoteStore(object):
+            def __init__(self):
+                for name, attr in attrs.items():
+                    setattr(self, name, attr)
+
+        return RemoteStore()
+
+    def _remote_attrs(self, run_id, dataset_idx):
+        return {
+            'train_data_path': self.get_train_data_path(dataset_idx),
+            'val_data_path': self.get_val_data_path(dataset_idx),
+            'test_data_path': self.get_test_data_path(dataset_idx),
+            'saving_runs': self.saving_runs(),
+            'runs_path': self.get_runs_path(),
+            'run_path': self.get_run_path(run_id),
+            'checkpoint_path': self.get_checkpoint_path(run_id),
+            'logs_path': self.get_logs_path(run_id),
+            'checkpoint_filename': self.get_checkpoint_filename(),
+            'logs_subdir': self.get_logs_subdir(),
+            'get_local_output_dir': self.get_local_output_dir_fn(run_id),
+            'sync': self.sync_fn(run_id)
+        }
+
+    @staticmethod
+    def create(prefix_path):
+        if HDFSStore.matches(prefix_path):
+            return HDFSStore(prefix_path)
+        else:
+            return LocalStore(prefix_path)
+
+
+class PrefixStore(Store):
+    def __init__(self, prefix_path, train_path=None, val_path=None, test_path=None, runs_path=None, save_runs=True):
+        self.prefix_path = self.get_full_path(prefix_path)
+        self._train_path = train_path or self._get_path('intermediate_train_data')
+        self._val_path = val_path or self._get_path('intermediate_val_data')
+        self._test_path = test_path or self._get_path('intermediate_test_data')
+        self._runs_path = runs_path or self._get_path('runs')
+        self._save_runs = save_runs
+        super(PrefixStore, self).__init__()
+
+    def exists(self, path):
+        return self.get_filesystem().exists(self.get_localized_path(path))
```

Comment:
I have test this locally, it reported errors. So here and following should be full path?

Reply:
Hey @ConeyLiu, what was the error, and what was the path that was passed in?  Using full paths do not work with the PyArrow `HadoopFilesystem` class.

Reply:
I think I see where your error is coming from.  I'm assuming you're using full URLs like `hdfs://namenode01:8020/user/test/horovod`.  I updated the `HDFSStore` to work with these strings for the `prefix_path`, let me know if it solves your issue.

Reply:
sorry for the late reply. It works great.